File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        final List<List<Integer>> result = new ArrayList<>();
        visit(root, result, 0);
        return result;
    }

    private void visit(TreeNode node, List<List<Integer>> result, int treeLevel) {
        if (node == null) {
            return;
        }

        if (result.size() <= treeLevel) {
            result.add(new ArrayList<>());
        }

        result.get(treeLevel).add(node.val);

        visit(node.left, result, treeLevel + 1);
        visit(node.right, result, treeLevel + 1);
    } 
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        curr_p = 0
        inorder_pos = {val: index for index, val in enumerate(inorder)}
        def visit(l: int, r: int) -> TreeNode:
            nonlocal curr_p
            if l > r:
                return None
            val = preorder[curr_p]
            curr_p += 1
            val_pos = inorder_pos[val]
            return TreeNode(val, visit(l, val_pos - 1), visit(val_pos + 1, r))

        return visit(0, len(preorder) - 1)
